### ISSUE TITLE
Chinese messages appear garbled in debug console

### ISSUE THREAD

**Issue by isidorn** [OmniSharp/omnisharp-vscode#1775]
As reported by @DFVSQY here https://github.com/Microsoft/vscode/issues/35235 
1. Have the following program

```C#
using System;

namespace files
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("文件 编辑 选择 查看 转到 调试 任务 帮助");
            Console.WriteLine("哈哈");
        }
    }
}
```

2. Have your windows machine setup to use simplified chineese
3. Have vscode setup to use simplified chineese
4. Run the program, notice garbled characters
5. Notice garbled output in the C# extension output channel

### Work around:
Add the following at the start of the program:
```C#
    Console.OutputEncoding = System.Text.Encoding.UTF8;
```

### Alternative work around:
Edit launch.json to change `"console"` to use the integrated terminal:
```json
    "console": "integratedTerminal"
```

### Notes:
* this is not happening with other debug adapters thus I am assuming that the C# debug backend is sending wrong characters to vscode.
*  I was not able to reproduce this with just setting vscode to use chineese, I think the whole OS language has to be changed. 
* I get lots of bugs around this in the vscode repo from chineese people every mileston only for C# debugging.

Please ping if you need more details.

![31177450-a84ab022-a948-11e7-99c7-6be320405fa6](https://user-images.githubusercontent.com/1926584/31179969-a7a2e714-a91d-11e7-8412-dcb14171e717.gif)

**Comment by DustinCampbell** [comment]
@isidorn: This is a VS Code issue: https://github.com/Microsoft/vscode/issues/3550.

**Comment by gregg-miskelly** [comment]
@DFVSQY what is your `Console.OutputEncoding` set to in your program? (You can evaluate the property under the debugger to find out).

If it isn't `System.Text.UTF8Encoding`, does the issue go away if you explicitly set it to that?

**Comment by DFVSQY** [comment]
@gregg-miskelly  It seems like that the console depends on system encoding, if i change console.outputencoding to utf8, debug console is ok but the output window is still messy code, the default encoding is System.Text.OSEncoding,like below:
![test](https://user-images.githubusercontent.com/8075421/31204368-21e36c30-a99e-11e7-8852-ad051076a192.gif)
In addition, i'm a simplified chinese user, simplified chinese win10 and simplified chinese vscode.

**Comment by gregg-miskelly** [comment]
@DFVSQY Thanks. In this case, I think there are multiple bugs but with the same root cause -- non-UTF8 characters written to tools that expect UTF8 because on Windows there aren't great conversions for specifying output encoding. I will reopen this issue to track seeing if there is anything in the debugger that we can do to try and convince the implementation of Console.OutputEncoding to default to UTF-8.

For the build, by chance does it work if you add `/consoleloggerparameters:Encoding=UTF-8`?

**Comment by gregg-miskelly** [comment]
Debugger implementation notes: I believe this is where CoreFX works out the default console output page: https://github.com/dotnet/corefx/blob/7ce0acee7eaadc2f3f9776a76c166b29724c8acb/src/System.Console/src/System/ConsolePal.Windows.cs#L104

So we may be able to solve this by calling `SetConsoleCP(CP_UTF8)`.

**Comment by isidorn** [comment]
@gregg-miskelly thanks for reopening. The issue in vscode which @DustinCampbell pointed to is task specifc and is related to code pages and how windows determines printable characters in the terminal. This is however not related to this issue - unprintable characters in the output and debug console since those characters are coming directly from your extension.

**Comment by DFVSQY** [comment]
@gregg-miskelly  where i should put the /consoleloggerparameters:Encoding=UTF-8 in ? please give me an example, thank you.
![default](https://user-images.githubusercontent.com/8075421/31228112-e11a374e-aa0e-11e7-9dbc-1bf910f78afa.PNG)

**Comment by DustinCampbell** [comment]
@isidorn: The non-debug output example above is simply the VS Code Output pane running the build task.  C# for VS Code doesn't control that.

**Comment by isidorn** [comment]
@DustinCampbell yes the output is a dupilcate of the issue you linked. Thanks.

**Comment by gregg-miskelly** [comment]
@DFVSQY try setting `tasks[0].args` to `[ "${workspaceRoot}/files.csproj", "/consoleloggerparameters:Encoding=UTF-8" ]`. I am not sure if that will work or if we need something else to try and convince the .NET CLI to output UTF-8 text.

**Comment by DFVSQY** [comment]
@gregg-miskelly  i have a try for that parameter, but there is no difference in output window, it's still messy code.

**Comment by gregg-miskelly** [comment]
@DFVSQY Thanks for trying. One last idea:

```json
{
    "version": "0.1.0",
    "tasks": [
        {
            "taskName": "build",
            "command": "cmd",
            "suppressTaskName": true,
            "args": [
                "/c chcp 65001 >nul && dotnet build ${workspaceRoot}/files.csproj"
            ],
            "isBuildCommand": true,
            "problemMatcher": "$msCompile"
        }
    ]
}
```

**Comment by DFVSQY** [comment]
@gregg-miskelly i have a try for last idea, now it's ok, like below:
![test](https://user-images.githubusercontent.com/8075421/31258174-ffc52d0a-aa6f-11e7-9664-17976b49bab9.gif)

**Comment by gregg-miskelly** [comment]
Excellent. 

@dbaeumer: there might be a pretty trivial fix for https://github.com/Microsoft/vscode/issues/3550 by calling `SetConsoleCP(CP_UTF8)` in the vscode process that spawns task processes.

**Comment by dbaeumer** [comment]
@gregg-miskelly thanks. With the support of running tasks in the terminal this is not needed anymore since the terminal itself already allows to specify encodings.

**Comment by gregg-miskelly** [comment]
@bayesrule I will see if we can fix this in our next release. In the mean time, use the work around at the top of this issue.

**Comment by Svetomech** [comment]
Have the same problem (ru locale).

![image](https://user-images.githubusercontent.com/7114815/32634774-6002baac-c5be-11e7-9688-f329372be848.png)

**Comment by DustinCampbell** [comment]
This should be fixed when we moving test to be a task. cc @TheRealPiotrP

**Comment by dbaeumer** [comment]
@Svetomech run your tests using task version 2.0.0. Then they execute in a terminal for which either the correct code page is already set or you can set it using chcp.

**Comment by DFVSQY** [comment]
@dbaeumer yes, i have a try for the new task version, The terminal output is ok but if i use the **Console** default encoding, the debug console output is still messy code.
![test](https://user-images.githubusercontent.com/8075421/32658867-00f97bba-c657-11e7-8258-8b59f1f96a65.gif)

**Comment by Svetomech** [comment]
@dbaeumer how'd I do so?
This is what `dotnet new xunit` command leaves me with:
![image](https://user-images.githubusercontent.com/7114815/32666299-06721746-c648-11e7-9b9c-4ca08e60190a.png)
P.S. `dotnet --version` is 2.0.2

**Comment by DustinCampbell** [comment]
I think @dbaeumer assumed you had a custom task to run this. In this case, C# for VS Code is creating this particular window, so this is a bug on us to fix.

**Comment by Svetomech** [comment]
@DustinCampbell sounds not really related to this issue. I mean, I gave running tests as an example, but in actuality _all_ output in the Output window is encoded incorrectly. E.g. just pressing F5 results in this:
![image](https://user-images.githubusercontent.com/7114815/32666498-a971b4f6-c648-11e7-84a1-ad901529cb49.png)

**Comment by DustinCampbell** [comment]
@Svetomech : From an implementation perspective build and test are very different in C# for VS Code. The key difference is that the build task is defined in your project (in tasks.json) and can be easily converted to to tasks version 2.0. Then it will run in the Terminal as @dbaeumer. Testing with the "run/debug test" Code Lenses is not a task that is defined in your project. In that case, C# for VS Code runs specific code to handle this (not tasks) that output to a specific output channel created by the extension.

**Comment by Svetomech** [comment]
@DustinCampbell So, as I understand, the current workaround is to always prefer `"console": "integratedTerminal"` over `"console": "internalConsole"`? Can I set it up globally or do I have to edit **launch.json** on a per project basis?

**Comment by DustinCampbell** [comment]
I'm not sure whether that can be set globally. My goal in here was to point out that running tests with C# for VS Code is different and the workarounds in this issue might not help.

**Comment by gregg-miskelly** [comment]
@Svetomech : You can't set it globally.

I have updated the title of this bug to hopefully reflect the fact that this bug is only tracking fixing output in the debug console. If there are other items that need to be fixed lets get separate issues opened.

**Comment by gregg-miskelly** [comment]
@DustinCampbell do we need a similar issue for dealing with build command output for unit tests?

**Comment by DustinCampbell** [comment]
Yes, I think so. However, I think it is also related to @TheRealPiotrP's work on the task provider API. @TheRealPiotrP?

**Comment by gregg-miskelly** [comment]
@TheRealPiotrP do you want me to open another issue for unit tests? Or do you have this covered already?

**Comment by gregg-miskelly** [comment]
Fixed in 1.14.0-beta2

**Comment by DongchengWang** [comment]
I don't think this bug is fixed. Check out https://github.com/OmniSharp/omnisharp-vscode/issues/3437
